File: Output Languages/C Sharp - OHM/Base Library/Source/BaseTable.cs
/*
	Copyright 2009 KISS Intelligent Systems Limited
	
	   Licensed under the Apache License, Version 2.0 (the "License");
	   you may not use this file except in compliance with the License.
	   You may obtain a copy of the License at
	
	       http://www.apache.org/licenses/LICENSE-2.0
	
	   Unless required by applicable law or agreed to in writing, software
	   distributed under the License is distributed on an "AS IS" BASIS,
	   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	   See the License for the specific language governing permissions and
	   limitations under the License.
	   
*/

using System;
using System.Collections.Generic;
using System.Text;

using Apache.Hadoop.Hbase.Thrift;

namespace OHM
{

	public abstract class BaseTable
	{
		
		public BaseTable()
		{
		}

		public abstract void save();

		public abstract void delete();

		public abstract void loadAll();

		
		protected String generateNewKey(OHMConnection connection, byte[] tableName, Random rand)
		{
			String key = null;
			OHMRow result = null;
			
			bool found = false;

			while(!found)
			{			
				//Generate throw random 5 digit numbers
				int no1 = rand.Next(10000, 99999);
				int no2 = rand.Next(10000, 99999);
				int no3 = rand.Next(10000, 99999);

				key = no1.ToString() + no2.ToString() + no3.ToString();

				Console.WriteLine("New Key: " + key);
				
				try
				{
					 result = connection.getRow(tableName, fromString(key));
				}
				catch(Exception exp)
				{

				}

				if(result == null || result.Columns.Count == 0)
				{
					//The key is not already in use
					found = true;
				}
					
			}

			return key;
		}

		
		
		public static byte[] fromBool(bool value)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.GetBytes(value);
		}

		public static byte[] fromByte(byte value)
		{
			//Just put it straight in to the array
			return new byte[]{value};
		}

		public static byte[] fromByteStream(byte[] value)
		{
			//Just user the array directly
			return value;
		}
		
		public static byte[] fromDouble(double value)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.GetBytes(value);
		}

		public static byte[] fromFloat(float value)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.GetBytes(value);
		}
		
		public static byte[] fromInt(int value)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.GetBytes(value);
		}

		public static byte[] fromLong(long value)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.GetBytes(value);
		}

		public static byte[] fromString(String value)
		{
			//Use the UTF8 encoding
			return Encoding.UTF8.GetBytes(value);
		}

		public static byte[] fromUInt(uint value)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.GetBytes(value);
		}

		public static byte[] fromULong(ulong value)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.GetBytes(value);
		}

		
		public static bool toBool(byte[] data)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.ToBoolean(data, 0);
		}

		public static byte toByte(byte[] data)
		{
			if(data.Length > 0)
				//Just use the first value from the array
				return data[0];

			else
				return 0;
		}

		public static byte[] toByteStream(byte[] data)
		{
			//Just use the array directly
			return data;
		}

		public static double toDouble(byte[] data)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.ToDouble(data, 0);
		}

		public static float toFloat(byte[] data)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.ToSingle(data, 0);
		}
		
		public static int toInt(byte[] data)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.ToInt32(data, 0);
		}

		public static long toLong(byte[] data)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.ToInt64(data, 0);
		}
		
		public static String toString(byte[] data)
		{
			//Use the UTF8 encoding
			return Encoding.UTF8.GetString(data);
		}
		
		public static uint toUInt(byte[] data)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.ToUInt32(data, 0);
		}

		public static ulong toULong(byte[] data)
		{
			//Use Bit Conveters Little Endian conversion
			return BitConverter.ToUInt64(data, 0);
		}

		
	}
}

File: Output Languages/Java/Base Library/src/com/kissintelligentsystems/ocm/java/BaseTable.java
package com.kissintelligentsystems.ocm.java;

import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.util.Random;

import me.prettyprint.cassandra.service.Keyspace;

import org.apache.thrift.TException;

public abstract class BaseTable 
{
	public abstract String getKey();
	
	public abstract void newKey(Random rand) throws Exception;
	
	public abstract void save() throws Exception;

	public abstract void delete() throws Exception;

	public abstract void loadAll() throws Exception;

	
	protected String generateNewKey(OCMConnection connection, byte[] tableName, Random rand) throws IllegalArgumentException, org.apache.cassandra.thrift.NotFoundException, TException, Exception
	{
		Keyspace keyspace = null;
		String key = null;
		try
		{
			keyspace = connection.borrowKeySpace();
			

						
			//Generate a new random 5 digit numbers
			int no1 = rand.nextInt();
			int no2 = rand.nextInt();
			int no3 = rand.nextInt();
			
			key = Integer.toString(no1) +  Integer.toString(no2) + Integer.toString(no3);	
		}
		finally
		{
			connection.returnKeySpace(keyspace);
		}
		
		return key;
	}
	
	
	public static byte[] fromBoolean(boolean value)
	{	
		byte[] result = new byte[1];
		
		if(value)
			result[0] = 1; 
		
		else
			result[0] = 0; 
		
		return result;
	}

	public static byte[] fromByte(byte value)
	{
		//Just put it straight in to the array
		return new byte[]{value};
	}

	public static byte[] fromByteStream(byte[] value)
	{
		//Just user the array directly
		return value;
	}
	
	public static byte[] fromDouble(double value)
	{
		ByteBuffer bb = ByteBuffer.allocate(4); 
		bb.putDouble(value); 
		return bb.array();
	}

	public static byte[] fromFloat(float value)
	{
		ByteBuffer bb = ByteBuffer.allocate(4); 
		bb.putFloat(value); 
		return bb.array();
	}
	
	public static byte[] fromInt(int value)
	{
		ByteBuffer bb = ByteBuffer.allocate(4); 
		bb.putInt(value); 
		return bb.array();
	}

	public static byte[] fromLong(long value)
	{
		ByteBuffer bb = ByteBuffer.allocate(4); 
		bb.putLong(value); 
		return bb.array();
	}

	public static byte[] fromString(String value)
	{
		try
		{
			//Use the UTF8 encoding
			return value.getBytes("UTF8");
			
		} catch (UnsupportedEncodingException e)
		{
			return new byte[0];
		}
	}

	
	public static boolean toBoolean(byte[] data)
	{
		if(data.length == 1)
		{
			if(data[0] == 1)
				return true;
			
			else
				return false;
		}	
		
		return false;
	}

	public static byte toByte(byte[] data)
	{
		if(data.length > 0)
			//Just use the first value from the array
			return data[0];

		else
			return 0;
	}

	public static byte[] toByteStream(byte[] data)
	{
		//Just use the array directly
		return data;
	}

	public static double toDouble(byte[] data)
	{
		ByteBuffer bb = ByteBuffer.wrap(data);
		return bb.getDouble();
	}

	public static float toFloat(byte[] data)
	{
		ByteBuffer bb = ByteBuffer.wrap(data);
		return bb.getFloat();
	}
	
	public static int toInt(byte[] data)
	{
		ByteBuffer bb = ByteBuffer.wrap(data);
		return bb.getInt();
	}

	public static long toLong(byte[] data)
	{
		ByteBuffer bb = ByteBuffer.wrap(data);
		return bb.getLong();
	}
	
	public static String toString(byte[] data)
	{
		try
		{
			//Use the UTF8 encoding

			return new String(data, "UTF8");
		}
		catch (UnsupportedEncodingException e)
		{
			return null;
		}
	}
}
